.github/workflows/pre-merge-checks.yaml

```diff
@@ -0,0 +1,27 @@
+name: Pre Merge Checks
+on:
+  push:
+    paths-ignore:
+      - diagrams/**
+      - README.md
+  pull_request:
+    paths-ignore:
+      - diagrams/**
+      - README.md
+
+jobs:
+  test:
+    name: Run Tests
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: docker/setup-buildx-action@v3
+      - uses: docker/build-push-action@v5
+        with:
+          context: .
+          tags: todo-app:test
+          target: test
+          outputs: type=docker
+          cache-from: type=gha
+          cache-to: type=gha
```

Comment:
i spent a while trying to get caching to work, but it ended up being slower (random runs are much faster though)

Reply:
Not sure I totally follow, so from what I can see:
* Initial runs took ~25 seconds, including ~20 seconds of build time ([e.g. this run](https://github.com/josephcwalker/DevOps-Course-Starter/actions/runs/9758901599/job/26934477660))
* With the caching change:
  * _If caching hits_, we can cut that down to ~15 seconds total ([e.g. this](https://github.com/josephcwalker/DevOps-Course-Starter/actions/runs/9761370678/job/26942251200))
  * _But if caching misses_, then it's substantially slower ([e.g. this](https://github.com/josephcwalker/DevOps-Course-Starter/actions/runs/9761566040/job/26942887679)) at ~50 seconds
  * And looking more closely, it's the exporting & caching the data after a miss that is the slow part.. 

So it would depend on our caching hit rate and overall build time as to whether this is worth it, I agree!

Reply:
(Also, in case it contains anything of interest by comparison, I explored this previously with a learner and ended up with [this repo](https://github.com/JackMead/docker-caching/blob/main/.github/workflows/caching-test.yml) exploring how to do it without the underlying action here)